### DISCUSSION TITLE
[9 Settembre 2022] Cache

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#60] [Cache (M-Z)]
![Esame](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/d75674a4-b9bb-44fc-abf2-3f5e4ea02f85)
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11666134/AE.2021-22.-.Appello.3.-.2022-09-09.-.Esercizio.cache.-.Scheda.di.calcolo.pdf)

**Comment by FeddyLix17** [comment]
## 1) Accessi in memoria

dopo essersi ricavati le formule usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/fd201146-0131-4645-8732-ba8a472bc45e)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/b13425fa-6a44-443b-be32-0f74bb4d2c1d)

## 2) Dimensioni delle cache

considerando che ogni linea della cache è formata

- dal bit di validità

- dal *Least Recently Used* bit (solo L1)

- dal tag

- dal blocco stesso

si ottiene, rispettivamente per le 2 cache
 
- $\Large \text{L1}$
  
   il numero delle linee è dato dal prodotto tra le vie e i set, da cui

$$
    \Large  (1  + 1 + 27 + 2 * 4 * 8) * (2 * 4) \implies (29 + 64) * 8 \implies  93 * 8 \implies 744\ \text{bit}
$$

- $\Large \text{L2}$

$$
    \Large  (1 + 23 + 32 * 4 * 8) * 4 \implies (24 + 1024) * 4 \implies  1048 * 4 \implies 4192\ \text{bit}
$$

## 3) Tempi della sequenza di accessi

### (a) Tempo totale sequenza di accessi

rielencandoli brevemente dalla foto superiore, si hanno

- 4 accessi in memoria, con tempo $\Large 400 * 4 = 1600\ \text{ns}$

- 1 accessi in L1, con tempo  $\Large 1 * 1 = 1\ \text{ns}$

- 7 accessi in L2, con tempo  $\Large 40 * 7 = 280\ \text{ns}$

$$
    \Large T_{tot} = 1600 + 1 + 280 = 1881\ \text{ns}
$$

### (b) Tempo medio sequenza di accessi

$$
    \Large T_{med} = \frac{\text{T.\ Totale}}{\text{Num.\ accessi}} = \frac{1881}{12} = 156.75\ \text{ns}
$$

### (c) Istruzioni svolte nel tempo medio

$$
    \Large \text{Istruzioni\ eseguite}  = \frac{\text{T.\ medio} *\ \text{Frequenza}}{\text{Clock\ per\ Istruction}} = \frac{156.75 * 4}{4} = 156.75
$$

## 4) Word offset indirizzo 192 Cache L2

$$
    \Large \text{Word Offset} = (\text{Address} \gg \text{Byte Offset})\ \\%\ 2^{\text{bit Word Offset}}
$$

$$
    \Large  (192 \gg 2)\ \\%\ 2^{5} \implies 48\ \\%\ 32 \implies 16
$$

## 5) Pagine Virtuali

ricavandosi la formula usando un approccio grafico

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/eda72754-4a49-4679-a8f5-ae0de6f695ba)

si riempie la tabella ottenendo

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/20d7030d-af13-490a-9465-9dfd5feed911)

**Reply by matypist** [reply]
Confermo la proposta di soluzione di @FeddyLix17 , tutto chiaro e corretto 👍

<details>
<summary>Nota sul punto 5 (ma non particolarmente rilevante)</summary>
*512 pagine, non 1024
</details>

**Comment by rimaout** [comment]
### 1)
<img width="1022" alt="90C47210-9D99-4C03-838E-D6B663E13728" src="https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/116072651/2883b99e-7b62-4503-b62d-9306d03c74ba">

### 2)
<img width="1042" alt="44BCAF6C-6B19-4809-876F-9BE11990B70F" src="https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/116072651/9f8a9583-0294-4383-a299-ec4929cf03ad">

### 3)
![IMG_3152](https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/116072651/7861b1e4-70ed-4746-9a1a-b48eb3621854)

### 4)
<img width="948" alt="F8A3734E-AE66-4FFE-8C12-3F98F6DD202B" src="https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/116072651/733c46aa-7118-44fa-a5a0-84a04eac6ff9">

### 5)
<img width="853" alt="0D2D9455-61FA-47BE-B332-A7A5C420C95E" src="https://github.com/sapienzastudentsnetwork/architettura-degli-elaboratori/assets/116072651/b98b51a2-c89f-4c5d-86c5-0f760b78eefa">

**Comment by M1SC0** [comment]
![image](https://github.com/user-attachments/assets/ce89759c-f136-4458-863e-6166c3eeb965)

**Comment by Nw4pS** [comment]
per l'esercizio 5 basta effettuare (indirizzo / dimensione di una pagina) e prenderne la parte intera, quindi mi chiedo... a cosa serve sapere che la memoria virtuale consti di 512 pagine?
